### DISCUSSION TITLE
Collecting Istio Metrics without Prometheus Collector

### DISCUSSION THREAD

**Discussion by isu-kim** [istio/istio#49433] [Q&A] [answered]
Hello everyone, I have been looking for some ways to collect Istio metrics from our project.
However, as far as I know, Istio seems to export metrics to only Prometheus collectors. 

The following links were that I have tried:
- https://istio.io/latest/docs/reference/config/metrics/
- https://istio.io/latest/docs/tasks/observability/metrics/

I do understand that I can retrieve metrics from individual Envoy containers. However, since our goal is to retrieve the whole cluster's metrics from Istio, unfortunately, this is not suitable for our situation.

Also, in meshConfig, it seems like the `PrometheusMetricsProvider` for `ExtensionProvider` does not support custom services. Furthermore, as far as I can see with the actual [metrics source code from Istio](https://github.com/istio/istio/blob/master/istioctl/pkg/metrics/metrics.go#L123), 
```go
	// only use the first pod in the list
	promPod := pl.Items[0]
	fw, err := client.NewPortForwarder(promPod.Name, ctx.IstioNamespace(), "", 0, 9090)
	if err != nil {
		return fmt.Errorf("could not build port forwarder for prometheus: %v", err)
	}

```
it seems to have a fixed Prometheus collector (which would probably be `prometheus.istio-system.svc.cluster.local:9090`).

Am I missing something? or is there any other workaround for this issue?
(I am trying to avoid messing with Prometheus to forward data to our solution since this will require extra steps)
If so, any piece of information will be a huge help!
 
Also, if this discussion violates the guidelines, please let me know.
Thanks in advance for your help!

**Comment by zirain** [comment]
I think you can use custom bootstrap to configure stats sink.

**Comment by howardjohn** [comment, marked as answer]
I think you may be misunderstanding how Prometheus works, or I misunderstand the question. Istio does not _push_ metrics to prometheus; prometheus pulls metrics from the proxies. The proxies and Istio don't need to know, nor do they know, the address of prometheus; the flow of traffic is reverse. The code you linked is just `istioctl dashboard` helper to open up default prometheus UI -- its not relevant to metrics collection.

**Reply by isu-kim** [reply]
Thanks for your detailed answer! I was confused that Envoy _sends_ data to central Prometheus server defined in meshConfig, which the traffic was totally reversed. Now I get what I was missing, thank you again for your help!
